spark/src/test/scala/org/apache/spark/sql/delta/DeltaTypeWideningSuite.scala

```diff
@@ -0,0 +1,371 @@
+      if (partitioned) {
+        // The table needs to have at least 1 non-partition column, use a dummy one.
+        append(testCase.initialValuesDF.withColumn("dummy", lit(1)), partitionBy = Seq("value"))
+      } else {
+        append(testCase.initialValuesDF)
+      }
+      sql(s"ALTER TABLE delta.`$tempPath` " +
+        s"SET TBLPROPERTIES('delta.feature.timestampNtz' = 'supported')")
+
+      val alterTableSql =
+        s"ALTER TABLE delta.`$tempPath` CHANGE COLUMN value TYPE ${testCase.toType.sql}"
+      checkError(
+        exception = intercept[AnalysisException] {
+          sql(alterTableSql)
+        },
+        errorClass = "NOT_SUPPORTED_CHANGE_COLUMN",
+        sqlState = None,
+        parameters = Map(
+          "table" -> s"`spark_catalog`.`delta`.`$tempPath`",
+          "originName" -> toSQLId("value"),
+          "originType" -> toSQLType(testCase.fromType),
+          "newName" -> toSQLId("value"),
+          "newType" -> toSQLType(testCase.toType)),
+        context = ExpectedContext(
+          fragment = alterTableSql,
+          start = 0,
+          stop = alterTableSql.length - 1)
+      )
+    }
+  }
+
+  test("type widening in nested fields") {
+    sql(s"CREATE TABLE delta.`$tempPath` " +
+      "(s struct<a: byte>, m map<byte, short>, a array<short>) USING DELTA")
+    append(Seq((1, 2, 3, 4))
+      .toDF("a", "b", "c", "d")
+      .selectExpr(
+        "named_struct('a', cast(a as byte)) as s",
+        "map(cast(b as byte), cast(c as short)) as m",
+        "array(cast(d as short)) as a"))
+
+    assert(readDeltaTable(tempPath).schema === new StructType()
+      .add("s", new StructType().add("a", ByteType))
+      .add("m", MapType(ByteType, ShortType))
+      .add("a", ArrayType(ShortType)))
+
+    sql(s"ALTER TABLE delta.`$tempPath` CHANGE COLUMN s.a TYPE short")
+    sql(s"ALTER TABLE delta.`$tempPath` CHANGE COLUMN m.key TYPE int")
+    sql(s"ALTER TABLE delta.`$tempPath` CHANGE COLUMN m.value TYPE int")
+    sql(s"ALTER TABLE delta.`$tempPath` CHANGE COLUMN a.element TYPE int")
+
+    assert(readDeltaTable(tempPath).schema === new StructType()
+      .add("s", new StructType().add("a", ShortType))
+      .add("m", MapType(IntegerType, IntegerType))
+      .add("a", ArrayType(IntegerType)))
+
+    append(Seq((5, 6, 7, 8))
+      .toDF("a", "b", "c", "d")
+        .selectExpr("named_struct('a', cast(a as short)) as s", "map(b, c) as m", "array(d) as a"))
+
+    checkAnswer(
+      readDeltaTable(tempPath),
+      Seq((1, 2, 3, 4), (5, 6, 7, 8))
+        .toDF("a", "b", "c", "d")
+        .selectExpr("named_struct('a', cast(a as short)) as s", "map(b, c) as m", "array(d) as a"))
+  }
+
+  test("type widening using ALTER TABLE REPLACE COLUMNS") {
+    append(Seq(1, 2).toDF("value").select($"value".cast(ShortType)))
+    assert(readDeltaTable(tempPath).schema === new StructType().add("value", ShortType))
+    sql(s"ALTER TABLE delta.`$tempPath` REPLACE COLUMNS (value INT)")
```

Comment:
Same here -- what if we're replacing an entire STRUCT column and widening one of the fields inside the struct? Or a MAP or an ARRAY?

Reply:
Yes, that's actually a pretty important test case since replacing top-level columns is the only way you can change nested type with REPLACE COLUMNS. I added a couple of tests for this